### DISCUSSION TITLE
some questions about the deploy procudure of the project

### DISCUSSION THREAD

**Discussion by maodia** [eclipse/jkube#2342] [Q&A]
1. i am using resource fragments now , but i do not know how to customize the resource name. For example ,  i edit the resource fragment file service.yml,  add the servicename like:

```yml
metadata:
     name: my-service
```

then i excute 'mvn k8s:resource',  i will get twe service yaml files , one named '${project-name}-service.yml' and the other named 'my-service-service.yml' . Next  if  i excute 'mvn k8s:deploy' , it will create two services in the k8s cluster , that is not what i expect . i only expect the  service  with the custom name to be created in the k8s cluster.  

2.   after excuting the goal of k8s:build , can i skip the goal of k8s:push and direct deploy the local docker image to the k8s cluster by excuting the goal of k8s:deploy or k8s:apply?

3.  is there any quick way to switch which k8s cluster to deploy to?   we need this if we have two or more k8s clusters.

that are all my questions , expect your answer, thanks.

**Comment by manusa** [comment]
### Service Fragments

To change the name of the service, the recommended approach is to override the
[`jkube.enricher.jkube-name.name`](https://eclipse.dev/jkube/docs/kubernetes-maven-plugin/#jkube-name), or the more specific [`jkube.enricher.jkube-service.name`](https://eclipse.dev/jkube/docs/kubernetes-maven-plugin/#jkube-service) properties.

The problem with the fragment approach is that we need to provide the means for users to add additional services. The only way to achieve this is while still being able to customize the default generated service is to name the service fragment differently for the extra-service. This means that if you provide a service fragment with no name override, then those settings will be merged with the generated service. If however, you provide a service fragment with a different name, this fragment will be accounted for as an extra-service.

### Local image

You need to be able to push the image to the cluster's Kubernetes container image registry.
For example in Minikube you can run `eval $(minikube docker-env)` prior to building the image (`k8s:build`) and then you can directly generate the resources and apply them without worrying to push the image.

### Cluster configuration

JKube uses the Fabric8 Kubernetes Client to perform the deploy operations. The Kubernetes Client reads your cluster information from your [`.kube/config`](https://blog.marcnuri.com/where-is-my-default-kubeconfig-file).
You can either switch the context on your kube config `kubectl config use-context...`

If you don't want to rely on automatic configuration, then you can also provide your cluster configuration https://eclipse.dev/jkube/docs/kubernetes-maven-plugin/#_kubernetes_access_configuration

Maybe use 2 different Maven profiles for this so you can switch your cluster just by providing a `-P` maven flag.

**Reply by manusa** [reply]
By the way, these are really good questions that might be useful to others. If possible, next time try to open separate discussions for each so that we facilitate the Google indexing and GitHub query experience :pray:

**Reply by maodia** [reply]
@manusa 
ok，i will separate my questions next time, each of them will have its own discussion , if it is more helpful to others.

Thanks  for  your advices and detailed description， i will try your advices as soon as possible.

**Comment by maodia** [comment]
@manusa 
i tried to override the
 jkube.enricher.jkube-name.name or the jkube.enricher.jkube-service.name properties separately . The first one really changed the name of the deployment name and the service name in the yaml file , and the second did not change the name of service name in the yaml file.

**Reply by manusa** [reply]
Have you tried the `jkube.enricher.jkube-service.name` in isolation? If you set both, then one will take precendence over the other.
In case you've done so, have you removed the old fragment?
Is this a Spring Project and you've set the `spring.application.name` application property?

**Reply by maodia** [reply]
yes,i  tried the jkube.enricher.jkube-service.name in isolation and i have only one service.yml in the src/main/java/jkube folder without metadata.name property in it. Do you mean i should remove the service resource fragment file and set  all of the properties of service in the jkube.enricher.jkube-service tag?  it is a Spring project and i have not set the spring.application.name application property. i think the default name comes from the 'artifactId'  in my pom file.

**Reply by manusa** [reply]
Everything looks fine.
Could you provide a minimal reproducer project? So we can further check and try to understand what's going on

**Comment by maodia** [comment]
i have uploaded the minimal project. we can  excute   'mvn k8s:resorce'  which can reproduce the problem , tks.
[demo4k8service.zip](https://github.com/eclipse/jkube/files/12438407/demo4k8service.zip)

**Reply by manusa** [reply]
OK, I can reproduce it.

Indeed there's an issue while providing the Service Fragment.

I've tried to find a workaround, but it doesn't seem possible at the moment.

**Reply by maodia** [reply]
It's not that urgent,please take your time and happy weekend.

**Reply by manusa** [reply]
I noticed it's not linked to the discussion. I opened issue #2350 to fix it. You can track its progress there.
